### PR TITLE
fix(beacon-node): cleanly exit process on graceful shutdown

### LINKED ISSUE #5317: Beacon node exits uncontrolled (too early) during graceful shutdown
**Problem**

Lodestar BN does not wait for [all jobs](https://github.com/ChainSafe/lodestar/blob/d360e62bde4abb566444d71f45fdc4fd74584ae7/packages/beacon-node/src/node/nodejs.ts#L309) to be finished during graceful shutdown.
This is problematic as some operations are not executed such as writing BLSToExecutionChanges to disk (see [discord discussion](https://discord.com/channels/593655374469660673/593655641445367808/1090250663344209983))

**Root cause**

Lodestar uses [threads package](https://www.npmjs.com/package/@chainsafe/threads) which installs [SIGTERM signal listeners](https://github.com/andywer/threads.js/blob/0d1a882f09273cb8482902ae2601fe4ffed6fbab/src/master/implementation.node.ts#L153)  that exit the main process once all workers are terminated, see [terminateWorkersAndMaster](https://github.com/andywer/threads.js/blob/0d1a882f09273cb8482902ae2601fe4ffed6fbab/src/master/implementation.node.ts#L142).

**Related**

- https://github.com/andywer/threads.js/pull/198
- https://github.com/andywer/threads.js/issues/388
- https://github.com/andywer/threads.js/pull/324
- https://github.com/MatrixAI/Polykey-Desktop/issues/79
- https://github.com/andywer/threads.js/pull/329

### PR DESCRIPTION
**Pull request by nflaig**
**Motivation**

Cleanly exit beacon node on graceful shutdown without requiring `process.exit`, closes https://github.com/ChainSafe/lodestar/issues/5317

**Description**

Uses [why-is-node-running](https://github.com/mafintosh/why-is-node-running) to identify active handles. (alternativ tool: [wtfnode](https://github.com/myndzi/wtfnode))

~~Most of the time process exits but there are errors such as `Error: QUEUE_ERROR_QUEUE_ABORTED`.
In some cases the process keeps running, I guess depends on the timing...good news is that workers are all closed correctly~~

<details>
  <summary>why-is-node-running output</summary>
  <p>
There are 1411 handle(s) keeping the process running

# TTYWRAP
node:internal/async_hooks:202
node:internal/bootstrap/switches/is_main_thread:51
node:internal/bootstrap/switches/is_main_thread:170
/home/nico/projects/ethereum/lodestar/node_modules/siginfo/index.js:3               - if (process.stderr.isTTY || force === true) {
/home/nico/projects/ethereum/lodestar/node_modules/why-is-node-running/include.js:3 - require('siginfo')(why, true)
node:internal/modules/cjs/loader:1159
node:internal/modules/cjs/loader:1213
node:internal/modules/cjs/loader:1037

# FILEHANDLE
node:internal/async_hooks:202

# FILEHANDLE
node:internal/async_hooks:202

# FILEHANDLE
node:internal/async_hooks:202

... more node:internal/async_hooks:202

# TTYWRAP
node:internal/async_hooks:202
node:internal/bootstrap/switches/is_main_thread:51
node:internal/bootstrap/switches/is_main_thread:148
node:internal/bootstrap/loaders:189
node:internal/bootstrap/loaders:311
node:internal/bootstrap/loaders:291
node:internal/bootstrap/loaders:296

# TTYWRAP
node:internal/async_hooks:202
node:internal/bootstrap/switches/is_main_thread:196
node:internal/bootstrap/loaders:189
node:internal/bootstrap/loaders:311
node:internal/bootstrap/loaders:291
node:internal/bootstrap/loaders:296
node:internal/bootstrap/loaders:276

# DNSCHANNEL
node:internal/async_hooks:202
node:internal/dns/utils:79
node:internal/dns/utils:74
node:internal/dns/utils:241
node:internal/dns/utils:251
node:internal/dns/promises:367
node:internal/bootstrap/loaders:331
node:internal/bootstrap/loaders:362
node:dns/promises:3

# JSSTREAM
node:internal/async_hooks:202
node:internal/js_stream_socket:49
/home/nico/projects/ethereum/lodestar/node_modules/http2-wrapper/source/utils/js-stream-socket.js:6 - const JSStreamSocket = (new tls.TLSSocket(new stream.PassThrough()))._handle._parentWrap.constructor;
node:internal/modules/cjs/loader:1159
node:internal/modules/cjs/loader:1213
node:internal/modules/cjs/loader:1037
node:internal/modules/cjs/loader:878
node:internal/modules/cjs/loader:1061

# TLSWRAP
node:internal/async_hooks:202
/home/nico/projects/ethereum/lodestar/node_modules/http2-wrapper/source/utils/js-stream-socket.js:6 - const JSStreamSocket = (new tls.TLSSocket(new stream.PassThrough()))._handle._parentWrap.constructor;
node:internal/modules/cjs/loader:1159
node:internal/modules/cjs/loader:1213
node:internal/modules/cjs/loader:1037
node:internal/modules/cjs/loader:878
node:internal/modules/cjs/loader:1061

# FILEHANDLE
node:internal/async_hooks:202

# FILEHANDLE
node:internal/async_hooks:202

# FILEHANDLE
node:internal/async_hooks:202

# FILEHANDLE
node:internal/async_hooks:202

... more node:internal/async_hooks:202

# ZLIB
node:internal/async_hooks:202
/home/nico/projects/ethereum/lodestar/node_modules/got/node_modules/decompress-response/index.js:43 - const decompressStream = isBrotli ? zlib.createBrotliDecompress() : zlib.createUnzip();
file:///home/nico/projects/ethereum/lodestar/node_modules/got/dist/source/core/index.js:549
file:///home/nico/projects/ethereum/lodestar/node_modules/got/dist/source/core/index.js:736
file:///home/nico/projects/ethereum/lodestar/node_modules/got/dist/source/core/index.js:754

# nan:test.DelayRequest
node:internal/async_hooks:202
/home/nico/projects/ethereum/lodestar/node_modules/gc-stats/index.js:17                           - gcstats.afterGC(function(stats) {
/home/nico/projects/ethereum/lodestar/node_modules/prometheus-gc-stats/index.js:58                - gc().on('stats', stats => {
file:///home/nico/projects/ethereum/lodestar/packages/beacon-node/lib/metrics/nodeJsMetrics.js:14
file:///home/nico/projects/ethereum/lodestar/packages/beacon-node/lib/metrics/metrics.js:21
file:///home/nico/projects/ethereum/lodestar/packages/beacon-node/lib/node/nodejs.js:81
node:internal/process/task_queues:95
file:///home/nico/projects/ethereum/lodestar/packages/cli/lib/cmds/beacon/handler.js:52

# ZLIB
node:internal/async_hooks:202
/home/nico/projects/ethereum/lodestar/node_modules/node-fetch/lib/index.js:1622 - body = body.pipe(zlib.createGunzip(zlibOptions));

# ZLIB
node:internal/async_hooks:202
/home/nico/projects/ethereum/lodestar/node_modules/node-fetch/lib/index.js:1622 - body = body.pipe(zlib.createGunzip(zlibOptions));

# ZLIB
node:internal/async_hooks:202
/home/nico/projects/ethereum/lodestar/node_modules/node-fetch/lib/index.js:1622 - body = body.pipe(zlib.createGunzip(zlibOptions));

# ZLIB
node:internal/async_hooks:202
/home/nico/projects/ethereum/lodestar/node_modules/node-fetch/lib/index.js:1622 - body = body.pipe(zlib.createGunzip(zlibOptions));

# ZLIB
node:internal/async_hooks:202
/home/nico/projects/ethereum/lodestar/node_modules/node-fetch/lib/index.js:1622 - body = body.pipe(zlib.createGunzip(zlibOptions));

# ZLIB
node:internal/async_hooks:202
/home/nico/projects/ethereum/lodestar/node_modules/node-fetch/lib/index.js:1622 - body = body.pipe(zlib.createGunzip(zlibOptions));

# ZLIB
node:internal/async_hooks:202
/home/nico/projects/ethereum/lodestar/node_modules/node-fetch/lib/index.js:1622 - body = body.pipe(zlib.createGunzip(zlibOptions));

# ZLIB
node:internal/async_hooks:202
/home/nico/projects/ethereum/lodestar/node_modules/node-fetch/lib/index.js:1622 - body = body.pipe(zlib.createGunzip(zlibOptions));

# ZLIB
node:internal/async_hooks:202
/home/nico/projects/ethereum/lodestar/node_modules/node-fetch/lib/index.js:1622 - body = body.pipe(zlib.createGunzip(zlibOptions));

# ZLIB
node:internal/async_hooks:202
/home/nico/projects/ethereum/lodestar/node_modules/node-fetch/lib/index.js:1622 - body = body.pipe(zlib.createGunzip(zlibOptions));

# ZLIB
node:internal/async_hooks:202
/home/nico/projects/ethereum/lodestar/node_modules/node-fetch/lib/index.js:1622 - body = body.pipe(zlib.createGunzip(zlibOptions));

# ZLIB
node:internal/async_hooks:202
/home/nico/projects/ethereum/lodestar/node_modules/node-fetch/lib/index.js:1622 - body = body.pipe(zlib.createGunzip(zlibOptions));

# ZLIB
node:internal/async_hooks:202
/home/nico/projects/ethereum/lodestar/node_modules/node-fetch/lib/index.js:1622 - body = body.pipe(zlib.createGunzip(zlibOptions));

# ZLIB
node:internal/async_hooks:202
/home/nico/projects/ethereum/lodestar/node_modules/node-fetch/lib/index.js:1622 - body = body.pipe(zlib.createGunzip(zlibOptions));

# ZLIB
node:internal/async_hooks:202
/home/nico/projects/ethereum/lodestar/node_modules/node-fetch/lib/index.js:1622 - body = body.pipe(zlib.createGunzip(zlibOptions));

# ZLIB
node:internal/async_hooks:202
/home/nico/projects/ethereum/lodestar/node_modules/node-fetch/lib/index.js:1622 - body = body.pipe(zlib.createGunzip(zlibOptions));

# ZLIB
node:internal/async_hooks:202
/home/nico/projects/ethereum/lodestar/node_modules/node-fetch/lib/index.js:1622 - body = body.pipe(zlib.createGunzip(zlibOptions));

# ZLIB
node:internal/async_hooks:202
/home/nico/projects/ethereum/lodestar/node_modules/node-fetch/lib/index.js:1622 - body = body.pipe(zlib.createGunzip(zlibOptions));

# ZLIB
node:internal/async_hooks:202
/home/nico/projects/ethereum/lodestar/node_modules/node-fetch/lib/index.js:1622 - body = body.pipe(zlib.createGunzip(zlibOptions));

# ZLIB
node:internal/async_hooks:202
/home/nico/projects/ethereum/lodestar/node_modules/node-fetch/lib/index.js:1622 - body = body.pipe(zlib.createGunzip(zlibOptions));

# ZLIB
node:internal/async_hooks:202
/home/nico/projects/ethereum/lodestar/node_modules/node-fetch/lib/index.js:1622 - body = body.pipe(zlib.createGunzip(zlibOptions));

# ZLIB
node:internal/async_hooks:202
/home/nico/projects/ethereum/lodestar/node_modules/node-fetch/lib/index.js:1622 - body = body.pipe(zlib.createGunzip(zlibOptions));

# ZLIB
node:internal/async_hooks:202
/home/nico/projects/ethereum/lodestar/node_modules/node-fetch/lib/index.js:1622 - body = body.pipe(zlib.createGunzip(zlibOptions));

# ZLIB
node:internal/async_hooks:202
/home/nico/projects/ethereum/lodestar/node_modules/node-fetch/lib/index.js:1622 - body = body.pipe(zlib.createGunzip(zlibOptions));

# ZLIB
node:internal/async_hooks:202
/home/nico/projects/ethereum/lodestar/node_modules/node-fetch/lib/index.js:1622 - body = body.pipe(zlib.createGunzip(zlibOptions));

# Timeout
node:internal/async_hooks:202
node:internal/async_hooks:505
node:internal/timers:159
node:internal/timers:193
file:///home/nico/projects/ethereum/lodestar/packages/beacon-node/lib/chain/clock/LocalClock.js:23
node:internal/timers:564
node:internal/timers:507

# ZLIB
node:internal/async_hooks:202
/home/nico/projects/ethereum/lodestar/node_modules/node-fetch/lib/index.js:1622 - body = body.pipe(zlib.createGunzip(zlibOptions));

 </p>
</details>

However, it is much better than on `unstable` branch which never exits

<details>
  <summary>why-is-node-running output on unstable</summary>
  <p>
There are 1436 handle(s) keeping the process running

# TTYWRAP
node:internal/async_hooks:202
node:internal/bootstrap/switches/is_main_thread:51
node:internal/bootstrap/switches/is_main_thread:170
/home/nico/projects/ethereum/lodestar/node_modules/siginfo/index.js:3               - if (process.stderr.isTTY || force === true) {
/home/nico/projects/ethereum/lodestar/node_modules/why-is-node-running/include.js:3 - require('siginfo')(why, true)
node:internal/modules/cjs/loader:1159
node:internal/modules/cjs/loader:1213
node:internal/modules/cjs/loader:1037

# FILEHANDLE
node:internal/async_hooks:202

# FILEHANDLE
node:internal/async_hooks:202

# FILEHANDLE
node:internal/async_hooks:202

# FILEHANDLE
node:internal/async_hooks:202

....more node:internal/async_hooks:202

# TTYWRAP
node:internal/async_hooks:202
node:internal/bootstrap/switches/is_main_thread:51
node:internal/bootstrap/switches/is_main_thread:148
node:internal/bootstrap/loaders:189
node:internal/bootstrap/loaders:311
node:internal/bootstrap/loaders:291
node:internal/bootstrap/loaders:296

# TTYWRAP
node:internal/async_hooks:202
node:internal/bootstrap/switches/is_main_thread:196
node:internal/bootstrap/loaders:189
node:internal/bootstrap/loaders:311
node:internal/bootstrap/loaders:291
node:internal/bootstrap/loaders:296
node:internal/bootstrap/loaders:276

# DNSCHANNEL
node:internal/async_hooks:202
node:internal/dns/utils:79
node:internal/dns/utils:74
node:internal/dns/utils:241
node:internal/dns/utils:251
node:internal/dns/promises:367
node:internal/bootstrap/loaders:331
node:internal/bootstrap/loaders:362
node:dns/promises:3

# JSSTREAM
node:internal/async_hooks:202
node:internal/js_stream_socket:49
/home/nico/projects/ethereum/lodestar/node_modules/http2-wrapper/source/utils/js-stream-socket.js:6 - const JSStreamSocket = (new tls.TLSSocket(new stream.PassThrough()))._handle._parentWrap.constructor;
node:internal/modules/cjs/loader:1159
node:internal/modules/cjs/loader:1213
node:internal/modules/cjs/loader:1037
node:internal/modules/cjs/loader:878
node:internal/modules/cjs/loader:1061

# TLSWRAP
node:internal/async_hooks:202
/home/nico/projects/ethereum/lodestar/node_modules/http2-wrapper/source/utils/js-stream-socket.js:6 - const JSStreamSocket = (new tls.TLSSocket(new stream.PassThrough()))._handle._parentWrap.constructor;
node:internal/modules/cjs/loader:1159
node:internal/modules/cjs/loader:1213
node:internal/modules/cjs/loader:1037
node:internal/modules/cjs/loader:878
node:internal/modules/cjs/loader:1061

# ZLIB
node:internal/async_hooks:202
/home/nico/projects/ethereum/lodestar/node_modules/got/node_modules/decompress-response/index.js:43 - const decompressStream = isBrotli ? zlib.createBrotliDecompress() : zlib.createUnzip();
file:///home/nico/projects/ethereum/lodestar/node_modules/got/dist/source/core/index.js:549
file:///home/nico/projects/ethereum/lodestar/node_modules/got/dist/source/core/index.js:736
file:///home/nico/projects/ethereum/lodestar/node_modules/got/dist/source/core/index.js:754

# nan:test.DelayRequest
node:internal/async_hooks:202
/home/nico/projects/ethereum/lodestar/node_modules/gc-stats/index.js:17                           - gcstats.afterGC(function(stats) {
/home/nico/projects/ethereum/lodestar/node_modules/prometheus-gc-stats/index.js:58                - gc().on('stats', stats => {
file:///home/nico/projects/ethereum/lodestar/packages/beacon-node/lib/metrics/nodeJsMetrics.js:14
file:///home/nico/projects/ethereum/lodestar/packages/beacon-node/lib/metrics/metrics.js:21
file:///home/nico/projects/ethereum/lodestar/packages/beacon-node/lib/node/nodejs.js:81
node:internal/process/task_queues:95
file:///home/nico/projects/ethereum/lodestar/packages/cli/lib/cmds/beacon/handler.js:52

# ZLIB
node:internal/async_hooks:202
/home/nico/projects/ethereum/lodestar/node_modules/node-fetch/lib/index.js:1622 - body = body.pipe(zlib.createGunzip(zlibOptions));

# Timeout
node:internal/async_hooks:202
node:internal/async_hooks:505
node:internal/timers:159
node:internal/timers:193
file:///home/nico/projects/ethereum/lodestar/packages/reqresp/lib/rate_limiter/ReqRespRateLimiter.js:59
file:///home/nico/projects/ethereum/lodestar/packages/reqresp/lib/ReqResp.js:70
file:///home/nico/projects/ethereum/lodestar/packages/beacon-node/lib/network/reqresp/ReqRespBeaconNode.js:46
file:///home/nico/projects/ethereum/lodestar/packages/beacon-node/lib/network/network.js:231

# ZLIB
node:internal/async_hooks:202
/home/nico/projects/ethereum/lodestar/node_modules/node-fetch/lib/index.js:1622 - body = body.pipe(zlib.createGunzip(zlibOptions));

# ZLIB
node:internal/async_hooks:202
/home/nico/projects/ethereum/lodestar/node_modules/node-fetch/lib/index.js:1622 - body = body.pipe(zlib.createGunzip(zlibOptions));

# ZLIB
node:internal/async_hooks:202
/home/nico/projects/ethereum/lodestar/node_modules/node-fetch/lib/index.js:1622 - body = body.pipe(zlib.createGunzip(zlibOptions));

# ZLIB
node:internal/async_hooks:202
/home/nico/projects/ethereum/lodestar/node_modules/node-fetch/lib/index.js:1622 - body = body.pipe(zlib.createGunzip(zlibOptions));

# ZLIB
node:internal/async_hooks:202
/home/nico/projects/ethereum/lodestar/node_modules/node-fetch/lib/index.js:1622 - body = body.pipe(zlib.createGunzip(zlibOptions));

# ZLIB
node:internal/async_hooks:202
/home/nico/projects/ethereum/lodestar/node_modules/node-fetch/lib/index.js:1622 - body = body.pipe(zlib.createGunzip(zlibOptions));

# ZLIB
node:internal/async_hooks:202
/home/nico/projects/ethereum/lodestar/node_modules/node-fetch/lib/index.js:1622 - body = body.pipe(zlib.createGunzip(zlibOptions));

# ZLIB
node:internal/async_hooks:202
/home/nico/projects/ethereum/lodestar/node_modules/node-fetch/lib/index.js:1622 - body = body.pipe(zlib.createGunzip(zlibOptions));

# ZLIB
node:internal/async_hooks:202
/home/nico/projects/ethereum/lodestar/node_modules/node-fetch/lib/index.js:1622 - body = body.pipe(zlib.createGunzip(zlibOptions));

# ZLIB
node:internal/async_hooks:202
/home/nico/projects/ethereum/lodestar/node_modules/node-fetch/lib/index.js:1622 - body = body.pipe(zlib.createGunzip(zlibOptions));

# ZLIB
node:internal/async_hooks:202
/home/nico/projects/ethereum/lodestar/node_modules/node-fetch/lib/index.js:1622 - body = body.pipe(zlib.createGunzip(zlibOptions));

# ZLIB
node:internal/async_hooks:202
/home/nico/projects/ethereum/lodestar/node_modules/node-fetch/lib/index.js:1622 - body = body.pipe(zlib.createGunzip(zlibOptions));

# ZLIB
node:internal/async_hooks:202
/home/nico/projects/ethereum/lodestar/node_modules/node-fetch/lib/index.js:1622 - body = body.pipe(zlib.createGunzip(zlibOptions));

# ZLIB
node:internal/async_hooks:202
/home/nico/projects/ethereum/lodestar/node_modules/node-fetch/lib/index.js:1622 - body = body.pipe(zlib.createGunzip(zlibOptions));

# ZLIB
node:internal/async_hooks:202
/home/nico/projects/ethereum/lodestar/node_modules/node-fetch/lib/index.js:1622 - body = body.pipe(zlib.createGunzip(zlibOptions));

# ZLIB
node:internal/async_hooks:202
/home/nico/projects/ethereum/lodestar/node_modules/node-fetch/lib/index.js:1622 - body = body.pipe(zlib.createGunzip(zlibOptions));

# ZLIB
node:internal/async_hooks:202
/home/nico/projects/ethereum/lodestar/node_modules/node-fetch/lib/index.js:1622 - body = body.pipe(zlib.createGunzip(zlibOptions));

# ZLIB
node:internal/async_hooks:202
/home/nico/projects/ethereum/lodestar/node_modules/node-fetch/lib/index.js:1622 - body = body.pipe(zlib.createGunzip(zlibOptions));

# ZLIB
node:internal/async_hooks:202
/home/nico/projects/ethereum/lodestar/node_modules/node-fetch/lib/index.js:1622 - body = body.pipe(zlib.createGunzip(zlibOptions));

# ZLIB
node:internal/async_hooks:202
/home/nico/projects/ethereum/lodestar/node_modules/node-fetch/lib/index.js:1622 - body = body.pipe(zlib.createGunzip(zlibOptions));

# ZLIB
node:internal/async_hooks:202
/home/nico/projects/ethereum/lodestar/node_modules/node-fetch/lib/index.js:1622 - body = body.pipe(zlib.createGunzip(zlibOptions));

# ZLIB
node:internal/async_hooks:202
/home/nico/projects/ethereum/lodestar/node_modules/node-fetch/lib/index.js:1622 - body = body.pipe(zlib.createGunzip(zlibOptions));

# ZLIB
node:internal/async_hooks:202
/home/nico/projects/ethereum/lodestar/node_modules/node-fetch/lib/index.js:1622 - body = body.pipe(zlib.createGunzip(zlibOptions));

# ZLIB
node:internal/async_hooks:202
/home/nico/projects/ethereum/lodestar/node_modules/node-fetch/lib/index.js:1622 - body = body.pipe(zlib.createGunzip(zlibOptions));

# ZLIB
node:internal/async_hooks:202
/home/nico/projects/ethereum/lodestar/node_modules/node-fetch/lib/index.js:1622 - body = body.pipe(zlib.createGunzip(zlibOptions));

# ZLIB
node:internal/async_hooks:202
/home/nico/projects/ethereum/lodestar/node_modules/node-fetch/lib/index.js:1622 - body = body.pipe(zlib.createGunzip(zlibOptions));

# ZLIB
node:internal/async_hooks:202
/home/nico/projects/ethereum/lodestar/node_modules/node-fetch/lib/index.js:1622 - body = body.pipe(zlib.createGunzip(zlibOptions));

# ZLIB
node:internal/async_hooks:202
/home/nico/projects/ethereum/lodestar/node_modules/node-fetch/lib/index.js:1622 - body = body.pipe(zlib.createGunzip(zlibOptions));

# ZLIB
node:internal/async_hooks:202
/home/nico/projects/ethereum/lodestar/node_modules/node-fetch/lib/index.js:1622 - body = body.pipe(zlib.createGunzip(zlibOptions));

# ZLIB
node:internal/async_hooks:202
/home/nico/projects/ethereum/lodestar/node_modules/node-fetch/lib/index.js:1622 - body = body.pipe(zlib.createGunzip(zlibOptions));

# ZLIB
node:internal/async_hooks:202
/home/nico/projects/ethereum/lodestar/node_modules/node-fetch/lib/index.js:1622 - body = body.pipe(zlib.createGunzip(zlibOptions));

# ZLIB
node:internal/async_hooks:202
/home/nico/projects/ethereum/lodestar/node_modules/node-fetch/lib/index.js:1622 - body = body.pipe(zlib.createGunzip(zlibOptions));

# ZLIB
node:internal/async_hooks:202
/home/nico/projects/ethereum/lodestar/node_modules/node-fetch/lib/index.js:1622 - body = body.pipe(zlib.createGunzip(zlibOptions));

# ZLIB
node:internal/async_hooks:202
/home/nico/projects/ethereum/lodestar/node_modules/node-fetch/lib/index.js:1622 - body = body.pipe(zlib.createGunzip(zlibOptions));

# ZLIB
node:internal/async_hooks:202
/home/nico/projects/ethereum/lodestar/node_modules/node-fetch/lib/index.js:1622 - body = body.pipe(zlib.createGunzip(zlibOptions));

# ZLIB
node:internal/async_hooks:202
/home/nico/projects/ethereum/lodestar/node_modules/node-fetch/lib/index.js:1622 - body = body.pipe(zlib.createGunzip(zlibOptions));

# ZLIB
node:internal/async_hooks:202
/home/nico/projects/ethereum/lodestar/node_modules/node-fetch/lib/index.js:1622 - body = body.pipe(zlib.createGunzip(zlibOptions));

# ZLIB
node:internal/async_hooks:202
/home/nico/projects/ethereum/lodestar/node_modules/node-fetch/lib/index.js:1622 - body = body.pipe(zlib.createGunzip(zlibOptions));

# ZLIB
node:internal/async_hooks:202
/home/nico/projects/ethereum/lodestar/node_modules/node-fetch/lib/index.js:1622 - body = body.pipe(zlib.createGunzip(zlibOptions));

# ZLIB
node:internal/async_hooks:202
/home/nico/projects/ethereum/lodestar/node_modules/node-fetch/lib/index.js:1622 - body = body.pipe(zlib.createGunzip(zlibOptions));

# ZLIB
node:internal/async_hooks:202
/home/nico/projects/ethereum/lodestar/node_modules/node-fetch/lib/index.js:1622 - body = body.pipe(zlib.createGunzip(zlibOptions));

# ZLIB
node:internal/async_hooks:202
/home/nico/projects/ethereum/lodestar/node_modules/node-fetch/lib/index.js:1622 - body = body.pipe(zlib.createGunzip(zlibOptions));

# ZLIB
node:internal/async_hooks:202
/home/nico/projects/ethereum/lodestar/node_modules/node-fetch/lib/index.js:1622 - body = body.pipe(zlib.createGunzip(zlibOptions));

# ZLIB
node:internal/async_hooks:202
/home/nico/projects/ethereum/lodestar/node_modules/node-fetch/lib/index.js:1622 - body = body.pipe(zlib.createGunzip(zlibOptions));

# ZLIB
node:internal/async_hooks:202
/home/nico/projects/ethereum/lodestar/node_modules/node-fetch/lib/index.js:1622 - body = body.pipe(zlib.createGunzip(zlibOptions));

# ZLIB
node:internal/async_hooks:202
/home/nico/projects/ethereum/lodestar/node_modules/node-fetch/lib/index.js:1622 - body = body.pipe(zlib.createGunzip(zlibOptions));

# ZLIB
node:internal/async_hooks:202
/home/nico/projects/ethereum/lodestar/node_modules/node-fetch/lib/index.js:1622 - body = body.pipe(zlib.createGunzip(zlibOptions));

# ZLIB
node:internal/async_hooks:202
/home/nico/projects/ethereum/lodestar/node_modules/node-fetch/lib/index.js:1622 - body = body.pipe(zlib.createGunzip(zlibOptions));

# ZLIB
node:internal/async_hooks:202
/home/nico/projects/ethereum/lodestar/node_modules/node-fetch/lib/index.js:1622 - body = body.pipe(zlib.createGunzip(zlibOptions));

# Timeout
node:internal/async_hooks:202
node:internal/async_hooks:505
node:internal/timers:159
node:internal/timers:193
file:///home/nico/projects/ethereum/lodestar/packages/beacon-node/lib/chain/clock/LocalClock.js:23
file:///home/nico/projects/ethereum/lodestar/packages/beacon-node/lib/chain/clock/LocalClock.js:37
file:///home/nico/projects/ethereum/lodestar/packages/beacon-node/lib/chain/clock/LocalClock.js:51
file:///home/nico/projects/ethereum/lodestar/packages/beacon-node/lib/network/network.js:328
</p>
</details>

**How to use this branch**

- Checkout branch
```sh
git checkout -b nflaig/cleanly-exit-bn origin/nflaig/cleanly-exit-bn
```

- Install & build
```sh
yarn install && yarn build
```

- Run beacon node
```sh
./lodestar beacon --dataDir <data_dir> --execution.urls <el_url> --jwt-secret <jwtsecret> --network goerli
```

- Get pid from log
```sh
Process ID <pid>
```

- Wait for BN to be started, at least sync should start

```log
info: Syncing - 2.5 hours left...
```

- Ctrl +C to send SIGINT to process

- Trigger why-is-node-running logs from another shell (see [instructions](https://github.com/mafintosh/why-is-node-running#cli))
```sh
kill -SIGUSR1 <pid>
```

**Next steps**

- [x] identify why process does not shutdown cleanly in some cases
- [x] get rid of `Error: QUEUE_ERROR_QUEUE_ABORTED` and other errors
- [x] implement stoppable `Eth1DepositDataTracker`
- [x] probably still need to call `process.exit` just to make sure...but lets clean up first

**Cleanup**

- [x] remove why-is-node-running
- [x] remove patches/@chainsafe+threads+1.10.0.patch and patch-package
- [x] remove console.log

### CODE REVIEW

File: packages/beacon-node/src/eth1/index.ts [hunk 1]
```diff
@@ -70,7 +70,7 @@ export class Eth1ForBlockProduction implements IEth1ForBlockProduction {
 
     this.eth1DepositDataTracker = opts.disableEth1DepositDataTracker
       ? null
-      : new Eth1DepositDataTracker(opts, modules, eth1Provider);
+      : new Eth1DepositDataTracker({...opts, enabled: false}, modules, eth1Provider);
```
Comment (nflaig) [inline packages/beacon-node/src/eth1/index.ts]: just disabled this for now, requires more code changes to correctly stop deposit data tracker

File: patches/@chainsafe+threads+1.10.0.patch
Comment (nflaig) [inline patches/@chainsafe+threads+1.10.0.patch:1]: patched via [patch-package](https://www.npmjs.com/package/patch-package) until new version of `@chainsafe/threads` is released

**Comment by nflaig** [comment]
quick fix to ensure BN process is stopped correctly (there is probably a cleaner way to do this)

```diff
diff --git a/packages/cli/src/cmds/beacon/handler.ts b/packages/cli/src/cmds/beacon/handler.ts
index cd640822c3..e63b93f7fe 100644
--- a/packages/cli/src/cmds/beacon/handler.ts
+++ b/packages/cli/src/cmds/beacon/handler.ts
@@ -92,7 +92,18 @@ export async function beaconHandler(args: BeaconArgs & GlobalArgs): Promise<void
       abortController.abort();
     }, logger.info.bind(logger));
 
-    abortController.signal.addEventListener("abort", () => node.close(), {once: true});
+    abortController.signal.addEventListener(
+      "abort",
+      () =>
+        node
+          .close()
+          .then(() => process.exit(0))
+          .catch((e) => {
+            logger.error("Error closing beacon node", {}, e);
+            process.exit(1);
+          }),
+      {once: true}
+    );
   } catch (e) {
     await db.stop();

```

However, might want to get rid of those errors as well

```log
CApr-01 14:03:07.118[]                 info: Stopping gracefully, use Ctrl+C again to force process exit
Apr-01 14:03:08.103[chain]           error: Failed to run prepareForNextSlot nextEpoch=166257, isEpochTransition=false, prepareSlot=5320216 Aborted
Error: Aborted
    at EventTarget.onAbort (file:///home/nico/projects/ethereum/lodestar/packages/utils/src/sleep.ts:24:14)
    at EventTarget.[nodejs.internal.kHybridDispatch] (node:internal/event_target:731:20)
    at EventTarget.dispatchEvent (node:internal/event_target:673:26)
    at abortSignal (node:internal/abort_controller:308:10)
    at AbortController.abort (node:internal/abort_controller:338:5)
    at BeaconChain.close (file:///home/nico/projects/ethereum/lodestar/packages/beacon-node/src/chain/chain.ts:278:26)
    at BeaconNode.close (file:///home/nico/projects/ethereum/lodestar/packages/beacon-node/src/node/nodejs.ts:322:24)
    at processTicksAndRejections (node:internal/process/task_queues:95:5)
```

Related https://github.com/ChainSafe/lodestar/pull/2141

File: packages/beacon-node/src/node/nodejs.ts [hunk 2]
```diff
@@ -323,6 +323,7 @@ export class BeaconNode {
       await this.db.stop();
       if (this.controller) this.controller.abort();
       this.status = BeaconNodeStatus.closed;
+      console.log("Beacon node closed!");
```
Comment (g11tech) [inline packages/beacon-node/src/node/nodejs.ts]: console.log (just to keep track)

File: packages/beacon-node/src/node/nodejs.ts [hunk 3]
```diff
@@ -74,6 +75,12 @@ enum LoggerModule {
   sync = "sync",
 }
 
+/**
+ * Short delay before closing db to give async operations sufficient time to finish
+ * and prevent "Database is not open" errors when shutting down beacon node.
+ */
+const DELAY_BEFORE_CLOSING_DB_MS = 1000;
```
Comment (nflaig) [inline packages/beacon-node/src/node/nodejs.ts]: ~~500ms seems sufficient in most cases but might be good to have some safety margin~~

After further testing 500ms seems sufficient

File: packages/beacon-node/src/node/nodejs.ts [hunk 4]
```diff
@@ -316,11 +323,11 @@ export class BeaconNode {
       if (this.metricsServer) await this.metricsServer.stop();
       if (this.monitoring) this.monitoring.stop();
       if (this.restApi) await this.restApi.close();
-
       await this.chain.persistToDisk();
       await this.chain.close();
-      await this.db.stop();
       if (this.controller) this.controller.abort();
```
Comment (nflaig) [inline packages/beacon-node/src/node/nodejs.ts:328]: Abort must be called before closing db

Comment (nflaig) [reply packages/beacon-node/src/node/nodejs.ts:328]: In general, it is best to abort as late as possible in shutdown sequence. This ensures all intervals/timeouts are cleared first and then all remaining pending operations are aborted. Prevents edge case where abort is done and interval schedules another operation which would be stuck pending and delay shutdown.

File: packages/beacon-node/src/node/nodejs.ts [hunk 5]
```diff
@@ -74,6 +75,12 @@ enum LoggerModule {
   sync = "sync",
 }
 
+/**
+ * Short delay before closing db to give async operations sufficient time to complete
+ * and prevent "Database is not open" errors when shutting down beacon node.
```
Comment (nflaig) [inline packages/beacon-node/src/node/nodejs.ts:80]: As per [documentation](https://github.com/Level/classic-level#dbclosecallback), "Database is not open" only happens if close is already called and some async operation still tries to execute db queries. If db query is already executed before, the database will wait for any pending operations to finish before closing.

**Review by wemeetagain — APPROVED** [review]

**Comment by wemeetagain** [comment]
:tada: This PR is included in [v1.9.0](https://github.com/ChainSafe/lodestar/releases/tag/v1.9.0) :tada:

**Pull request merged**